File: programmers/level2/컬러링북/컬러링북.java
import java.util.*;

class Solution {
    boolean[][] visit = new boolean[100][100];
    int dx[] = {1, -1, 0, 0};
    int dy[] = {0, 0, 1, -1};
    int N, M;
    
    public int bfs(int y, int x, int var, int[][] picture) {
        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(y, x));
        visit[y][x] = true;
        int count = 1;
        
        while(!q.isEmpty()) {
            int cy = q.peek().y;
            int cx = q.peek().x;
            int cvar = picture[cy][cx];
            q.remove();
            
            for (int i = 0; i < 4; i++) {
                int ny = cy + dy[i];
                int nx = cx + dx[i];
                
                if (ny < 0 || ny >= M || nx < 0 || nx >= N) continue;
                int nvar = picture[ny][nx];
                
                if (var == nvar && !visit[ny][nx]) {
                    q.add(new Pair(ny, nx));
                    visit[ny][nx] = true;
                    count++;
                }
                
            }
        }
        
        return count;
    }
    
    public int[] solution(int m, int n, int[][] picture) {
        int numberOfArea = 0;
        int maxSizeOfOneArea = 0;
        int[] answer = new int[2];
        N = n; M = m;
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (!visit[i][j] && picture[i][j] != 0) {
                    numberOfArea++;
                    maxSizeOfOneArea = Math.max(maxSizeOfOneArea, bfs(i, j, picture[i][j], picture));
                }
            }
        }
     
        answer[0] = numberOfArea;
        answer[1] = maxSizeOfOneArea;
        return answer;
    }
}

class Pair {
    int x;
    int y;
    
    public Pair(int y, int x) {
        this.x = x;
        this.y = y;
    }
}

File: programmers/level2/컬러링북/컬러링북.cpp
// week15-2 카카오프렌즈 컬러링북

#include <vector>
#include <queue>
#include <utility>
#include <iostream>
using namespace std;

vector<vector<bool>> isVisit;
vector<vector<int>> cpPicture;
int dx[4] = {1, -1, 0, 0};
int dy[4] = {0, 0, 1, -1};
int c, r;

int bfs(int i, int j) {
    int count = 1;
    queue<pair<int, int>> q;
    q.push(make_pair(i, j));
    isVisit[i][j] = true;
    
    while (!q.empty()) {
        int curY = q.front().first;
        int curX = q.front().second;
        int cur = cpPicture[curY][curX];
        q.pop();
        
        for (int i = 0; i < 4; i++) {
            int nextY = curY + dy[i];
            int nextX = curX + dx[i];

            if (nextY < 0 || nextY >= c || nextX < 0 || nextX >= r) continue;

            int next = cpPicture[nextY][nextX];
            if (!isVisit[nextY][nextX] && cur == next) {
                q.push(make_pair(nextY, nextX));
                isVisit[nextY][nextX] = true;
                count++;
            }
        }
    }
    //cout << "count = " << count << endl;
    return count;
}

// 전역 변수를 정의할 경우 함수 내에 초기화 코드를 꼭 작성해주세요.
vector<int> solution(int m, int n, vector<vector<int>> picture) {
    int number_of_area = 0;
    int max_size_of_one_area = 0;
    
    // isVisit 초기화
    for (int i = 0; i < picture.size(); i++) {
        vector<bool> tmp1;
        for (int j = 0; j < picture[0].size(); j++) {
            tmp1.push_back(false);
        }
        isVisit.push_back(tmp1);
    }
    
    // bfs
    for (int i = 0; i < picture.size(); i++) {
        for (int j = 0; j < picture[0].size(); j++) {
            if (!isVisit[i][j] && picture[i][j] != 0) {
                cout << "[" << i << "][" << j << "] ::: " << picture[i][j]; 
                max_size_of_one_area = max(max_size_of_one_area, bfs(i, j));
                number_of_area++;
            }
        }
    }
    
    vector<int> answer(2);
    answer[0] = number_of_area;
    answer[1] = max_size_of_one_area;
    
    return answer;
}

